### ISSUE TITLE
"Introduced in" and/or "Changed in" fields

### ISSUE THREAD

**Issue by Julian** [sourcemeta/learnjsonschema.com#20]
It might be nice to have a "introduced in" field for keywords which says what draft of the spec the keyword was last introduced or last changed behavior in.

I use this info a lot as part of writing tests or tooling and do it manually by flipping through the spec and migration docs. It might be too niche for 'normal users' or it might still be useful, putting it here to see what you think!

**Comment by jviotti** [comment]
I like the idea. I can dig into when each keyword was first defined, but I'm not sure I know very well when they were modified. 

Is the latter information you already have in the top of your head? If so and you write it here, I'll put it properly on the site

**Comment by jviotti** [comment]
I guess I can look it up in the breaking changes section, assuming those are complete

**Comment by jviotti** [comment]
OK, "Introduced In" implemented: https://github.com/sourcemeta/learnjsonschema.com/pull/29. It is my best take by re-reading all versions of the spec, but let me know if missed anything

**Comment by Julian** [comment]
(Just saw that somehow) -- will have a look.

**Comment by jviotti** [comment]
@Era-cell Given your current research, I think this is one that you can probably help a lot with. I did "Introduced In" for 2020-12 already, and it would be great to list all the dialects on which a specific keyword was changed in anyway.

For example, we know that `$id` was introduced in Draft 6 and was changed in 2019-09 to not support the anchor syntax anymore.  If you get this information (introduced in vs changed in) for every keyword in 2020-12 (for example, in our GitHub Projects or just as a comment here), I can quickly extend the site to show it

**Comment by jviotti** [comment]
@Era-cell Or in case you want to contribute them yourself, I created this quick WIP PR: https://github.com/Intelligence-AI/learnjsonschema.com/pull/225. It allows you to set the dialects a keyword was changed in using a `changed_in` property array. Feel free to take it over if you want

**Comment by jviotti** [comment]
Once this site has all of this information for every dialect, I'll do some nice charts like a lineage graph, a stability heat map, etc :)

**Comment by Era-cell** [comment]
Also, what if the keyword wasn't changed at all. Should I let it stay as it is?

**Comment by jviotti** [comment]
> Also, what if the keyword wasn't changed at all. I should not include changed_in attribute in the markdown. right?

Correct. Just omit `changed_in`.

> And, should I include changed_in if there is even a minute change.

Yes. If there was any change, you should include `changed_in`.

> And what about format keyword? It changes for every dialecct. So, is it good to have stacked up version values in the changed_in attribute?

Yes. Let's record `changed_in` for any dialect where `format` was changed. However, note we only have 2020-12 on the site right now! So it shouldn't be very hard. We'll consider other dialects when we add them to the site

**Comment by Era-cell** [comment]
Hi @jviotti , there are many kind/type of changes. Like:
Change/removal of location of where the keyword can appear, like $schema can appear in resource schema locations only
Change in values of the keyword, like $vocabualry should have "format-assertion", "annotation" instead of "format"
Change in how annotations are collected from th keyword, like "contains" annotations weren't collected before.
and maybe more.
So, I think going with value change and how value can be represented will be a good choice??

**Comment by jviotti** [comment]
@Era-cell 

> Change/removal of location of where the keyword can appear, like $schema can appear in resource schema locations only
> Change in how annotations are collected from th keyword, like "contains" annotations weren't collected before.

Yep, these are all valid changes worth capturing in `changed_in`!

> Change in values of the keyword, like $vocabualry should have "format-assertion", "annotation" instead of "format"

Though this is a vocabulary change, not a keyword change. We can omit this one.

> So, I think going with value change and how value can be represented will be a good choice??

Any of the semantic values you exemplified before (including keyword value changes!) are all worth documenting in the site! This site is only 2020-12 (for now!) so hopefully not very hard to documented "changed_in" here. I think most of the subtle changes you mentioned happen on other dialects.

**Comment by Era-cell** [comment]
Hey! So,  what I will have in `changed_in` attribute is 2020-12 or nothing. So, only changes from 2019 - 2020 should be captured?

**Comment by jviotti** [comment]
- If a keyword was added in 2020-12, don't put `changed_in`
- If a 2020-12 keyword was last changed in dialect X, then set `changed_in` to X. For example, if keyword `foo` was last changed in Draft 7, and remained the same in 2019-09, set `changed_in: draft7`
- If a 2020-12 keyword was changed various times across different dialects, set `changed_in` to the last dialect that is not 2020-12. For example, if keyword `foo` was changed in Draft 3, Draft 7, and 2019-09, just set `changed_in: 2019-09`

**Comment by Era-cell** [comment]
So, $id was changed in 2019 to not support anchor syntax. So, changed from draft 7. So, draft 7 should be included right?
So, the clarification is that "from" should be included in the change and for multiple changes only the recent change should be considered?
So for the draft PR (for $ref keyword): which has 
changed_in:
  - draft6
  - 2019-09
  We'll just include 2019-09 i.e, latest changed_in for now?

**Comment by jviotti** [comment]
> So, $id was changed in 2019 to not support anchor syntax. So, changed from draft 7. So, draft 7 should be included right?

We only document 2020-12 on learnjsonschema.com right now so we don't care about changes between older dialects yet.

> So for the draft PR (for $ref keyword): which has draft6, 2019-09, We'll just include 2019-09 i.e, latest changed_in for now?

Ah, I forgot I extended the PR to support changes of more than one dialect already. Let's document all changed dialects then!

**Comment by Era-cell** [comment]
Also, 
1. in draft6 schema's could be boolean, so even items, not can also be boolean. 
2. However for additonalProperties, the drafts from 0 always say "If false is provided, no additional properties are allowed, and the schema can not be extended". But explicit "boolean" mention is in draft3.
So, should all the keywords taking schema should be considered as changed? and what about 2?

**Comment by Era-cell** [comment]
> Draft 4 said that JSON Schemas are objects, but then in most cases that matters, actual keyword documentation said that they could be set to either objects or booleans. Then Draft 6 generalized this idea at the schema level. I think there are some cases where in theory there was a difference (like not being able to put booleans as items in i.e. `allOf`), but in practice nobody tends to do that.
> 
> > Also seeing contentEncoding. It was present in draft 0,1,2 was moved to Hyperschema in draft3. And brought back to draft7 later. So, nothing should be mentioned in changed_in right?
> 
> If the keyword just moved across vocabularies, then that's a vocabulary change, not a keyword change.
> 
> > Also I didn't understand what were the changes in $ref in draft6 and 2019-09
> 
> From the top of my head, the only difference I remember is that in 2019-09 and later, you can have `$ref` alongside other keywords. Before, `$ref` would dismiss any other sibling keyword: https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/main/tests/draft4/ref.json#L145-L177

@jviotti `contentEncoding` was completely removed in draft4 and again introduced in draft7, so...

**Comment by Era-cell** [comment]
> > Draft 4 said that JSON Schemas are objects, but then in most cases that matters, actual keyword documentation said that they could be set to either objects or booleans. Then Draft 6 generalized this idea at the schema level. I think there are some cases where in theory there was a difference (like not being able to put booleans as items in i.e. `allOf`), but in practice nobody tends to do that.
> > > Also seeing contentEncoding. It was present in draft 0,1,2 was moved to Hyperschema in draft3. And brought back to draft7 later. So, nothing should be mentioned in changed_in right?
> > 
> > 
> > If the keyword just moved across vocabularies, then that's a vocabulary change, not a keyword change.
> > > Also I didn't understand what were the changes in $ref in draft6 and 2019-09
> > 
> > 
> > From the top of my head, the only difference I remember is that in 2019-09 and later, you can have `$ref` alongside other keywords. Before, `$ref` would dismiss any other sibling keyword: https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/main/tests/draft4/ref.json#L145-L177
> 
> @jviotti `contentEncoding` was completely removed in draft4 and again introduced in draft7, so...

And also I am sure at some point someone will get this question - But id was present from draft1, even defs, so shouldn't it be mentioned that the keyword was renamed..?

**Comment by jviotti** [comment]
Fixed in https://github.com/Intelligence-AI/learnjsonschema.com/pull/238
